### PR TITLE
[Do not merge] [CPDNPQ-1078] [SPIKE] Implement geocoded local authority search

### PR DESCRIPTION
**Pull request by Lockyy**
### Context

Ticket: https://dfedigital.atlassian.net/browse/CPDNPQ-1078

This is not to be merged into NPQ, this code will be used in another repository that will be set up later. That repo will use this code for searching local authorities.

### Changes proposed in this pull request

This spike uses activerecord-postgis-adapter, geocoder, and rgeo to translate free text user location search into a list of the nearest local authorities.

The GeoJSON file available from [the GOV Open Geography Portal](https://geoportal.statistics.gov.uk/datasets/ons::local-authority-districts-december-2022-boundaries-uk-bfc/explore) is loaded into the DB by `Services::GeojsonLoader` using the PostGIS adapter provided by `activerecord-postgis-adapter`. 

Geocoder is used to convert entered user searches into lon/lat coordinates.

Rgeo is then used to transform those lon/lat coordinates into [EPSG:27700 British National Grid projection](https://epsg.io/27700-1197) coordinates. These coordinates once transformed line up to the projection stored in the database which is also the 27700 projection. 

The transformed coordinates are then used in a distance query using <-> to find the nearest local authority geometries.

Note that the database adaptor had to be swapped over from postgres to postgis to enable this functionality.

### CODE REVIEW

File: app/lib/services/geojson_loader.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    def self.reload_geojsons(geojson_path)
+      empty_database
+      load_geojsons(geojson_path)
+    end
+
+    def self.empty_database
+      GeoLocalAuthority.delete_all
+    end
+
+    def self.read_file(geojson_path)
```
Comment (slawosz) [inline app/lib/services/geojson_loader.rb:12]: Would move this method below `load_geojsons` method as its helper.

File: app/lib/services/geojson_loader.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+module Services
+  class GeojsonLoader
+    def self.reload_geojsons(geojson_path)
+      empty_database
+      load_geojsons(geojson_path)
+    end
+
+    def self.empty_database
+      GeoLocalAuthority.delete_all
```
Comment (slawosz) [inline app/lib/services/geojson_loader.rb:9]: I would consider inlining instead of separate method.

File: app/lib/services/geojson_loader.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      file = File.read(geojson_path)
+      features = RGeo::GeoJSON.decode(file, json_parser: :json)
+    end
+
+    def self.load_geojsons(geojson_path)
+      Rails.logger.debug("Loading #{geojson_path}")
+
+      features = read_file(geojson_path)
+
+      features.each do |feature|
```
Comment (slawosz) [inline app/lib/services/geojson_loader.rb:22]: Would it make sense to wrap it into transaction?

File: app/models/geo_local_authority.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+class GeoLocalAuthority < ApplicationRecord
+  include RGeo::ActiveRecord::GeometryMixin
+
+  SRID = 27_700
+
+  def self.nearest_three_to(string)
```
Comment (slawosz) [inline app/models/geo_local_authority.rb:6]: Maybe `self.nearest_to(string, amount: 3)` instead hardcoding argument in method name?

File: app/models/geo_local_authority.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    location = Geocoder.search(string).first
+    by_distance(location.longitude.to_f, location.latitude.to_f).limit(3)
+  end
+
+  def self.by_distance(longitude, latitude)
```
Comment (slawosz) [inline app/models/geo_local_authority.rb:11]: How many results does it returns by default? Could it turn out to be expensive as do extensive search?

Comment (BroiSatse) [reply app/models/geo_local_authority.rb:11]: It doesn't really return any results - it returns AR scope. The scope itslef uses some of the postgis magic which is rathere quite performant. I wouldn't expect any delays based on the size of the database in this service

**Pull request closed without merging**